### PR TITLE
feature: complete the details of StartPodSandbox

### PR DESCRIPTION
**Pull request by Starnop**
Signed-off-by: Starnop <starnop@163.com>

### Ⅰ. Describe what this PR did
Provide an interface to start a sandbox pod which was stopped by accident and setup the network with network plugin.

**Note:** Only if the CNI-plugin supports returning the same IP with the same parameters of `containdID` ,`pod name`,`pod namespace`, we will make sure it reacquire the original network configuration  of sandbox, like IP address.

### Ⅱ. Does this pull request fix one issue?

None.

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

None.

### Ⅳ. Describe how to verify it
use [cri-tools](https://github.com/Starnop/cri-tools/tree/cri-tools-ali)
```
# cat pod-config.json 
{
    "metadata": {
        "name": "nginx-sandbox",
        "namespace": "default",
        "attempt": 1,
        "uid": "hdishd83djaidwnduwk28bcsb"
    },
    "linux": {
    }
}

# cat container-config.json 
{
  "metadata": {
      "name": "busybox"
  },
  "image":{
      "image": "busybox"
  },
  "command": [
      "top"
  ],
  "linux": {
  }
}

# crictl runp pod-config.json 
DEBU[0000] RunPodSandboxRequest: &RunPodSandboxRequest{Config:&PodSandboxConfig{Metadata:&PodSandboxMetadata{Name:nginx-sandbox,Uid:hdishd83djaidwnduwk28bcsb,Namespace:default,Attempt:1,},Hostname:,LogDirectory:,DnsConfig:nil,PortMappings:[],Labels:map[string]string{},Annotations:map[string]string{},Linux:&LinuxPodSandboxConfig{CgroupParent:,SecurityContext:nil,Sysctls:map[string]string{},},},} 
DEBU[0000] RunPodSandboxResponse: &RunPodSandboxResponse{PodSandboxId:7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6,} 
7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6

# crictl create 7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6 container-config.json pod-config.json 
DEBU[0000] CreateContainerRequest: &CreateContainerRequest{PodSandboxId:7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6,Config:&ContainerConfig{Metadata:&ContainerMetadata{Name:busybox,Attempt:0,},Image:&ImageSpec{Image:busybox,},Command:[top],Args:[],WorkingDir:,Envs:[],Mounts:[],Devices:[],Labels:map[string]string{},Annotations:map[string]string{},LogPath:,Stdin:false,StdinOnce:false,Tty:false,Linux:&LinuxContainerConfig{Resources:nil,SecurityContext:nil,},Windows:nil,NetPriority:0,QuotaId:,},SandboxConfig:&PodSandboxConfig{Metadata:&PodSandboxMetadata{Name:nginx-sandbox,Uid:hdishd83djaidwnduwk28bcsb,Namespace:default,Attempt:1,},Hostname:,LogDirectory:,DnsConfig:nil,PortMappings:[],Labels:map[string]string{},Annotations:map[string]string{},Linux:&LinuxPodSandboxConfig{CgroupParent:,SecurityContext:nil,Sysctls:map[string]string{},},},} 
DEBU[0000] CreateContainerResponse: &CreateContainerResponse{ContainerId:a0bf0780844bb939c14cf43c2f34e42c99eb021fd35645dd1a1aeb6bbbce535f,} 
a0bf0780844bb939c14cf43c2f34e42c99eb021fd35645dd1a1aeb6bbbce535f

# crictl inspectp 7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6
DEBU[0000] PodSandboxStatusRequest: &PodSandboxStatusRequest{PodSandboxId:7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6,Verbose:true,} 
DEBU[0000] PodSandboxStatusResponse: &PodSandboxStatusResponse{Status:&PodSandboxStatus{Id:7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6,Metadata:&PodSandboxMetadata{Name:nginx-sandbox,Uid:hdishd83djaidwnduwk28bcsb,Namespace:default,Attempt:1,},State:SANDBOX_READY,CreatedAt:1536916075102934587,Network:&PodSandboxNetworkStatus{Ip:10.244.1.108,},Linux:&LinuxPodSandboxStatus{Namespaces:&Namespace{Options:&NamespaceOption{Network:POD,Pid:POD,Ipc:POD,},},},Labels:map[string]string{},Annotations:map[string]string{},},Info:map[string]string{},} 
{
  "status": {
    "id": "7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6",
    "metadata": {
      "attempt": 1,
      "name": "nginx-sandbox",
      "namespace": "default",
      "uid": "hdishd83djaidwnduwk28bcsb"
    },
    "state": "SANDBOX_READY",
    "createdAt": "2018-09-14T09:07:55.102934587Z",
    "network": {
      "ip": "10.244.1.108"
    },
    "linux": {
      "namespaces": {
        "options": {
          "ipc": "POD",
          "network": "POD",
          "pid": "POD"
        }
      }
    },
    "labels": {},
    "annotations": {}
  }
}

# crictl stopp 7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6
DEBU[0000] StopPodSandboxRequest: &StopPodSandboxRequest{PodSandboxId:7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6,} 
DEBU[0000] StopPodSandboxResponse: &StopPodSandboxResponse{} 
Stopped sandbox 7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6

# crictl startp 7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6
DEBU[0000] StartPodSandboxRequest: &StartPodSandboxRequest{PodSandboxId:7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6,} 
DEBU[0000] StartPodSandboxResponse: &StartPodSandboxResponse{} 
7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6

# crictl inspectp 7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6
DEBU[0000] PodSandboxStatusRequest: &PodSandboxStatusRequest{PodSandboxId:7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6,Verbose:true,} 
DEBU[0000] PodSandboxStatusResponse: &PodSandboxStatusResponse{Status:&PodSandboxStatus{Id:7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6,Metadata:&PodSandboxMetadata{Name:nginx-sandbox,Uid:hdishd83djaidwnduwk28bcsb,Namespace:default,Attempt:1,},State:SANDBOX_READY,CreatedAt:1536916075102934587,Network:&PodSandboxNetworkStatus{Ip:10.244.1.109,},Linux:&LinuxPodSandboxStatus{Namespaces:&Namespace{Options:&NamespaceOption{Network:POD,Pid:POD,Ipc:POD,},},},Labels:map[string]string{},Annotations:map[string]string{},},Info:map[string]string{},} 
{
  "status": {
    "id": "7fa3bf8c5cb8e97f1b09369bf182afb6b6f1ebdf74d072a95960942a3f2b9fb6",
    "metadata": {
      "attempt": 1,
      "name": "nginx-sandbox",
      "namespace": "default",
      "uid": "hdishd83djaidwnduwk28bcsb"
    },
    "state": "SANDBOX_READY",
    "createdAt": "2018-09-14T09:07:55.102934587Z",
    "network": {
      "ip": "10.244.1.108"
    },
    "linux": {
      "namespaces": {
        "options": {
          "ipc": "POD",
          "network": "POD",
          "pid": "POD"
        }
      }
    },
    "labels": {},
    "annotations": {}
  }
}

```

### Ⅴ. Special notes for reviews
Related Pull Request: https://github.com/alibaba/pouch/pull/2237

### CODE REVIEW

File: cri/v1alpha2/cri_utils.go [hunk 1]
```diff
@@ -448,6 +448,36 @@ func setupSandboxFiles(sandboxRootDir string, config *runtime.PodSandboxConfig)
 	return nil
 }
 
+// setupPodNetwork sets up the network of PodSandbox and return the netnsPath of PodSandbox
+// and do nothing when networkNamespaceMode equals runtime.NamespaceMode_NODE.
+func (c *CriManager) setupPodNetwork(ctx context.Context, id string, networkNamespaceMode runtime.NamespaceMode, config *runtime.PodSandboxConfig) (string, error) {
+	var netnsPath string
+	hostNet := networkNamespaceMode == runtime.NamespaceMode_NODE
+	// If it is in host network, no need to configure the network of sandbox.
+	if !hostNet {
```
Comment (allencloud) [inline cri/v1alpha2/cri_utils.go]: How about 
```
if hostNet {
    return netnsPath, nil
}
```

Use `return fast` to reduce ident and make code more clear.

Comment (Starnop) [reply cri/v1alpha2/cri_utils.go]: exactly, and has been updated. ^_^

**Comment by Starnop** [comment]
@rudyfly @YaoZengzeng PTAL. Thanks. ^_^

File: cri/v1alpha2/cri.go [hunk 2]
```diff
@@ -303,7 +282,42 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
...
+
+	// get the sandbox's meta data.
+	res, err := c.SandboxStore.Get(podSandboxID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get metadata of %q from SandboxStore: %v", podSandboxID, err)
+	}
+	sandboxMeta := res.(*SandboxMeta)
+
+	// Setup networking for the sandbox.
+	netnsPath, setupNeterr := c.setupPodNetwork(ctx, podSandboxID, sandboxMeta.Config.GetLinux().GetSecurityContext().GetNamespaceOptions().GetNetwork(), sandboxMeta.Config)
```
Comment (rudyfly) [inline cri/v1alpha2/cri.go]: s/setupNeterr/setupNetErr

Comment (Starnop) [reply cri/v1alpha2/cri.go]: Has been updated.

File: cri/v1alpha2/cri.go [hunk 3]
```diff
@@ -303,7 +282,42 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
...
+		return nil, fmt.Errorf("failed to get metadata of %q from SandboxStore: %v", podSandboxID, err)
+	}
+	sandboxMeta := res.(*SandboxMeta)
+
+	// Setup networking for the sandbox.
+	netnsPath, setupNetErr := c.setupPodNetwork(ctx, podSandboxID, sandboxMeta.Config.GetLinux().GetSecurityContext().GetNamespaceOptions().GetNetwork(), sandboxMeta.Config)
+	if setupNetErr != nil {
+		return nil, setupNetErr
+	}
+	defer func() {
```
Comment (YaoZengzeng) [inline cri/v1alpha2/cri.go]: If `setupNetErr` is not nil, the defer function will never being invoked.

File: cri/v1alpha2/cri.go [hunk 4]
```diff
@@ -303,7 +282,42 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 // and we should reconfigure it with network plugin which will make sure it reacquire its original network configuration,
 // like IP address.
 func (c *CriManager) StartPodSandbox(ctx context.Context, r *runtime.StartPodSandboxRequest) (*runtime.StartPodSandboxResponse, error) {
-	// TODO Complete the details of function.
+
+	podSandboxID := r.GetPodSandboxId()
+
+	// start Podsandbox.
```
Comment (YaoZengzeng) [inline cri/v1alpha2/cri.go]: Podsandbox -> PodSandbox

File: cri/v1alpha2/cri.go [hunk 5]
```diff
@@ -303,7 +282,42 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
...
+	}
+
+	// get the sandbox's meta data.
+	res, err := c.SandboxStore.Get(podSandboxID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get metadata of %q from SandboxStore: %v", podSandboxID, err)
+	}
+	sandboxMeta := res.(*SandboxMeta)
+
+	// Setup networking for the sandbox.
```
Comment (YaoZengzeng) [inline cri/v1alpha2/cri.go]: Setup -> setup

File: cri/v1alpha2/cri.go (hunk 3 above)
Comment (Starnop) [reply cri/v1alpha2/cri.go]: Yes, we should stop the container only if the network setting returns an error.

**Comment by YaoZengzeng** [comment]
LGTM

**Comment by rudyfly** [comment]
LGTM, and waiting for the ci passed.

**Pull request merged**